ledgerjs-fio/src/fio.ts

```diff
@@ -0,0 +1,335 @@
+
+const CLA = 0xd7
+
+function wrapConvertDeviceStatusError<T extends Function>(fn: T): T {
+    // @ts-ignore
+    return async (...args) => {
+        try {
+            return await fn(...args)
+        } catch (e: any) {
+            if (e && e.statusCode) {
+                throw new DeviceStatusError(e.statusCode)
+            }
+            throw e
+        }
+    }
+}
+
+/**
+ * FIO API
+ *
+ * @example
+ * import Fio from "@fioprotocol/fiojs";
+ * const fio = new Fio(transport);
+ */
+
+/** @ignore */
+export type SendFn = (params: SendParams) => Promise<Buffer>;
+
+// It can happen that we try to send a message to the device
+// when the device thinks it is still in a middle of previous ADPU stream.
+// This happens mostly if host does abort communication for some reason
+// leaving ledger mid-call.
+// In this case Ledger will respond by ERR_STILL_IN_CALL *and* resetting its state to
+// default. We can therefore transparently retry the request.
+
+// Note though that only the *first* request in an multi-APDU exchange should be retried.
+function wrapRetryStillInCall<T extends Function>(fn: T): T {
+    // @ts-ignore
+    return async (...args: any) => {
+        try {
+            return await fn(...args)
+        } catch (e: any) {
+            if (
+                e &&
+                e.statusCode &&
+                e.statusCode === DeviceStatusCodes.ERR_STILL_IN_CALL
+            ) {
+                // Do the retry
+                return await fn(...args)
+            }
+            throw e
+        }
+    }
+}
+
+
+async function interact<T>(
+    interaction: Interaction<T>,
+    send: SendFn,
+): Promise<T> {
+    let cursor = interaction.next()
+    let first = true
+    while (!cursor.done) {
+        const apdu = cursor.value
+        const res = first
+            ? await wrapRetryStillInCall(send)(apdu)
+            : await send(apdu)
+        first = false
+        cursor = interaction.next(res)
+    }
+    return cursor.value
+}
+
+/**
+ * Main API endpoint
+ * @category Main
+ */
+export class Fio {
+    /** @ignore */
+    transport: Transport<string>;
+    /** @ignore */
+    _send: SendFn;
+
+    constructor(transport: Transport<string>, scrambleKey: string = "FIO") {
+        this.transport = transport
+        // Note: this is list of methods that should "lock" the transport to avoid concurrent use
+        const methods = [
+            "getVersion",
+            "getSerial",
+            "getPublicKey",
+            "signTransaction",
+        ]
+        this.transport.decorateAppAPIMethods(this, methods, scrambleKey)
+        this._send = async (params: SendParams): Promise<Buffer> => {
+            let response = await wrapConvertDeviceStatusError(this.transport.send)(
+                CLA,
+                params.ins,
+                params.p1,
+                params.p2,
+                params.data,
+            )
+            response = utils.stripRetcodeFromResponse(response)
```

Comment:
Can be replaced with:
```suggestion
            response = stripRetcodeFromResponse(response)
```
and then the `import utils from "./utils"` replaced with: `import { stripRetcodeFromResponse } from './utils/ioHelpers'`